joern-cli/frontends/php2cpg/src/main/scala/io/joern/php2cpg/passes/PhpTypeRecovery.scala

```diff
@@ -0,0 +1,289 @@
+      case ::(identifier: Identifier, Nil) if symbolTable.contains(identifier) =>
+        symbolTable.get(identifier)
+      case ::(head: Call, Nil) =>
+        extractTypes(head.argument.l)
+      case _ => Set.empty
+    }
+    val returnTypes = extractTypes(ret.argumentOut.l)
+    existingTypes.addAll(returnTypes)
+
+    /* Check whether method return is already known, and if so, remove dummy value */
+    val saveTypes = existingTypes.filterNot(typeName => {
+      if (typeName.startsWith(Defines.UnresolvedNamespace))
+        true
+      else if (typeName.endsWith(s"${XTypeRecovery.DummyReturnType}"))
+        typeName.split(pathSep).headOption match {
+          case Some(methodName) => {
+            val methodReturns = methodReturnValues(Seq(methodName))
+              .filterNot(_.endsWith(s"${XTypeRecovery.DummyReturnType}"))
+            !methodReturns.isEmpty
+          }
+          case None => false
+        }
+      else
+        false
+    })
+    methodTypesTable.update(m, saveTypes)
+    builder.setNodeProperty(ret.method.methodReturn, PropertyNames.DYNAMIC_TYPE_HINT_FULL_NAME, saveTypes)
+  }
+
+  /* Necessary to change the filter regex from (this|self) to (\\$this|this), in order to account for $this PHP
+   * convention.
+   */
+  override protected def associateTypes(symbol: LocalVar, fa: FieldAccess, types: Set[String]): Set[String] = {
+    fa.astChildren.filterNot(_.code.matches("(\\$this|this|self)")).headOption.collect {
+      case fi: FieldIdentifier =>
+        getFieldParents(fa).foreach(t => persistMemberWithTypeDecl(t, fi.canonicalName, types))
+      case i: Identifier if isField(i) =>
+        getFieldParents(fa).foreach(t => persistMemberWithTypeDecl(t, i.name, types))
+    }
+    symbolTable.append(symbol, types)
+  }
+
+  /* Reference the PythonTypeRecovery implementation. The XTypeRecovery one seems incorrect. */
+  override protected def getFieldParents(fa: FieldAccess): Set[String] = {
+    if (fa.method.name == "<module>") {
+      Set(fa.method.fullName)
+    } else if (fa.method.typeDecl.nonEmpty) {
+      val parentTypes       = fa.method.typeDecl.fullName.toSet
+      val baseTypeFullNames = cpg.typeDecl.fullNameExact(parentTypes.toSeq: _*).inheritsFromTypeFullName.toSet
+      (parentTypes ++ baseTypeFullNames).filterNot(_.matches("(?i)(any|object)"))
+    } else {
+      super.getFieldParents(fa)
+    }
+  }
+
+  override protected def getTypesFromCall(c: Call): Set[String] = c.name match {
+    case Operators.fieldAccess        => symbolTable.get(LocalVar(getFieldName(new FieldAccess(c))))
+    case _ if symbolTable.contains(c) => symbolTable.get(c)
+    case Operators.indexAccess        => getIndexAccessTypes(c)
+    case n                            => methodReturnValues(Seq(c.methodFullName))
+  }
+
+  override protected def indexAccessToCollectionVar(c: Call): Option[CollectionVar] = {
+    def callName(x: Call) =
+      if (x.name.equals(Operators.fieldAccess))
```

Comment:
Scala is compare-by-value for Strings
```suggestion
      if (x.name == Operators.fieldAccess)
```

Reply:
I see a few other instances of this